teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+
+### Notes/Caveats
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for notes and caveats
+specific to supporting json results.
+
+* What if a Pipeline tries to use a object key that doesn't exist in a param or result?
+  * Some of this could be caught at creation time, i.e. pipeline params would declare the keys each object param will
+    contain, and we could validate at creation time that those keys exist
+  * For invalid uses that can only be caught at runtime, the PipelineRun would fail
+* What if a Task tries to write more keys to a result than the result declares it contains?
+  * The TaskRun would fail (alternatively we could decide to ignore the extra keys)
+* What if a Task tries to write less keys to a result than the result declares it contains?
+  * The TaskRun would fail
+
+### Risks and Mitigations
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for risks and mitigations
+specific to expanding ArrayOrString to support more types.
+
+### User Experience
+
+* Using the json schema syntax to describe the structure of the object is a bit verbose and might not be intuitive
+  to users; see [Alternatives](#alternatives) for discussion of an option where we create our own syntax.
+
+## Design Details
+
+### Declaring object results and params
+
+We would use [JSON object schema syntax](https://json-schema.org/understanding-json-schema/reference/object.html) to
+declare the structure of the object.
+
+Params example:
+
+```yaml
+  params:
+    - name: pull_remote
+      description: json schema has no "description" fields, so we'd have to include documentation about the structure in this field
+      schema:
+        type: object # see question below
+        properties:
+          url: {
+            type: string # for now, all values are strings, so we could imply it
+          }
+          path: {} # example of implying type: string
+```
+
+Results example:
+
+```yaml
+  results:
+    - name: sdist
+      description: json schema has no "description" fields, so we'd have to include documentation about the structure in this field
+      schema:
+        type: object
+        properties:
+          sha: {}
+          path: {}
+```
+
+**Question**: Do we want to keep using [the existing `type` field](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters)?
+i.e. do we use the top level `type: object` syntax, similar to how we declare `type: string` and `type: array` today, or
+do we want to move to always using `schema` and using the same syntax for arrays and strings as well (as suggested in
+[pipelines#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526))?
+
+**Proposal**: if possible (can find out with a proof of concept), we deprecated `type:`. We would still default to string
+when `schema` is not present, but when [array] or dict/object types are used, we use `schema`.
```

Comment:
currently i'm leaning toward continuing to support "type" and never actually deprecating it - ill add a section on this